### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by csmorley** [hussein-aitlahcen/FastNet#2]
Hi Hussein, thanks for the repo, seems to work well.

For a specific AbstractClient, is the data thread safe with one-worker deuquing per client?

I need to implement TCP framing, and need to know if i need to lock in the `server.EventStream.Subscribe` event handler, to ensure that data is not overwritten by a parallel thread for the same socket client.

Docs on the overview of the code would be awesome when you get a chance, or even links to techniques used would be great for learning - going through the code is using techniques (interesting) not seen before.

**Comment by hussein-aitlahcen** [comment]
Hello Chris,

This library is not thread-safe if you are planning to use a parallel type of IScheduler.
The **EventLoopScheduler** [being synchronous](https://github.com/dotnet/reactive/blob/master/Rx.NET/Source/src/System.Reactive/Concurrency/EventLoopScheduler.cs#L154) (used in my example), you are safe to **Subscribe** and consume events without locking. As you can see [here](https://github.com/hussein-aitlahcen/FastNet/blob/master/src/Core/AbstractServer.cs#L129), the **subscribed** handler is executed in the given Scheduler (EventLoopScheduler in our case).

To sum up, the underlying socket is asynchronous and produce network events that are handled in a synchronous manner.
The library should be patched so that you can't give a different Scheduler, and we should notice that it is tread-safe, a PR would be appreciated :).

I'm still here if you need more help !

**Comment by csmorley** [comment]
Awesome very helpful, will check reactive extensions. Next will check on
the memory buffer allocation, am implementing framing for tcp packets would
be nice to wrap on the standard buffer with one that implements the same
interface but looks for prefix byte count and then subsequent bytes in the
packet. So far server code is very fast and stable thank you. Will check
similair reactive constructs for the client side socket.
